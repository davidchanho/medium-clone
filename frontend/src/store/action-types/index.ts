export enum ActionTypes {
  FETCH_POSTS = "FETCH_POSTS",
  FETCH_POSTS_SUCCESS = "FETCH_POSTS_SUCCESS",
  FETCH_POSTS_FAIL = "FETCH_POSTS_FAIL",
  FETCH_POST = "FETCH_POST",
  FETCH_POST_SUCCESS = "FETCH_POST_SUCCESS",
  FETCH_POST_FAIL = "FETCH_POST_FAIL",
  ADD_POST = "ADD_POST",
  DELETE_POST = "DELETE_POST",
  UPDATE_POST = "UPDATE_POST",
}
