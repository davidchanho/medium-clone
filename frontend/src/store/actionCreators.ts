import { Dispatch } from "react";
import { Action, ActionTypes } from ".";
import { IComment, IPost, IUser } from "../types";
import API from "./api";

export const getComments = () => async (dispatch: Dispatch<Action>) => {
  dispatch({ type: ActionTypes.FETCH_COMMENTS });

  try {
    const data = await API.getComments();
    dispatch({
      type: ActionTypes.FETCH_COMMENTS_SUCCESS,
      payload: data,
    });
  } catch (err) {
    dispatch({
      type: ActionTypes.FETCH_COMMENTS_FAIL,
      payload: err,
    });
  }
};

export const getComment = (_id: string) => async (
  dispatch: Dispatch<Action>
) => {
  dispatch({ type: ActionTypes.FETCH_COMMENT });

  try {
    const data = await API.getComment(_id);
    dispatch({
      type: ActionTypes.FETCH_COMMENT_SUCCESS,
      payload: data,
    });
  } catch (err) {
    dispatch({
      type: ActionTypes.FETCH_COMMENT_FAIL,
      payload: err,
    });
  }
};

export const addComment = (comment: IComment) => async (
  dispatch: Dispatch<Action>
) => {
  try {
    API.addComment(comment);
    dispatch({ type: ActionTypes.ADD_COMMENT, payload: comment });
  } catch (err) {
    console.log(err);
  }
};

export const deleteComment = (_id: string) => async (
  dispatch: Dispatch<Action>
) => {
  try {
    API.deleteComment(_id);
    dispatch({
      type: ActionTypes.DELETE_COMMENT,
      payload: _id,
    });
  } catch (err) {
    console.log(err);
  }
};

export const updateComment = (comment: IComment) => async (
  dispatch: Dispatch<Action>
) => {
  try {
    API.updateComment(comment);
    dispatch({
      type: ActionTypes.UPDATE_COMMENT,
      payload: comment,
    });
  } catch (err) {
    console.log(err);
  }
};

export const getPosts = () => async (dispatch: Dispatch<Action>) => {
  dispatch({ type: ActionTypes.FETCH_POSTS });
  try {
    const data = await API.getPosts();

    dispatch({
      type: ActionTypes.FETCH_POSTS_SUCCESS,
      payload: data,
    });
  } catch (err) {
    dispatch({
      type: ActionTypes.FETCH_POSTS_FAIL,
      payload: err,
    });
  }
};

export const getPost = (_id: string) => async (dispatch: Dispatch<Action>) => {
  dispatch({ type: ActionTypes.FETCH_POST });

  try {
    const data = await API.getPost(_id);

    dispatch({
      type: ActionTypes.FETCH_POST_SUCCESS,
      payload: data,
    });
  } catch (err) {
    dispatch({
      type: ActionTypes.FETCH_POST_FAIL,
      payload: err,
    });
  }
};

export const addPost = (post: IPost) => async (dispatch: Dispatch<Action>) => {
  try {
    API.addPost(post);

    dispatch({ type: ActionTypes.ADD_POST, payload: post });
  } catch (err) {
    console.log(err);
  }
};

export const dismissPost = (_id: string) => async (
  dispatch: Dispatch<Action>
) => {
  try {
    dispatch({
      type: ActionTypes.DISMISS_POST,
      payload: _id,
    });
  } catch (err) {
    console.log(err);
  }
};

export const updatePost = (post: IPost) => async (
  dispatch: Dispatch<Action>
) => {
  try {
    API.updatePost(post);

    dispatch({
      type: ActionTypes.UPDATE_POST,
      payload: post,
    });
  } catch (err) {
    console.log(err);
  }
};

export const getPublications = () => async (dispatch: Dispatch<Action>) => {
  dispatch({ type: ActionTypes.FETCH_PUBLICATIONS });

  try {
    const data = await API.getPublications();
    dispatch({
      type: ActionTypes.FETCH_PUBLICATIONS_SUCCESS,
      payload: data,
    });
  } catch (err) {
    dispatch({
      type: ActionTypes.FETCH_PUBLICATIONS_FAIL,
      payload: err,
    });
  }
};

export const getPublication = (_id: string) => async (
  dispatch: Dispatch<Action>
) => {
  dispatch({ type: ActionTypes.FETCH_PUBLICATION });

  try {
    const data = await API.getPublication(_id);
    dispatch({
      type: ActionTypes.FETCH_PUBLICATION_SUCCESS,
      payload: data,
    });
  } catch (err) {
    dispatch({
      type: ActionTypes.FETCH_PUBLICATION_FAIL,
      payload: err,
    });
  }
};

export const getUsers = () => async (dispatch: Dispatch<Action>) => {
  dispatch({ type: ActionTypes.FETCH_USERS });

  try {
    const data = await API.getUsers();
    dispatch({
      type: ActionTypes.FETCH_USERS_SUCCESS,
      payload: data,
    });
  } catch (err) {
    dispatch({
      type: ActionTypes.FETCH_USERS_FAIL,
      payload: err,
    });
  }
};

export const getUser = (_id: string) => async (dispatch: Dispatch<Action>) => {
  dispatch({ type: ActionTypes.FETCH_USER });

  try {
    const data = await API.getUser(_id);
    dispatch({
      type: ActionTypes.FETCH_USER_SUCCESS,
      payload: data,
    });
  } catch (err) {
    dispatch({
      type: ActionTypes.FETCH_USER_FAIL,
      payload: err,
    });
  }
};

export const addUser = (user: IUser) => async (dispatch: Dispatch<Action>) => {
  try {
    API.addUser(user);
    dispatch({ type: ActionTypes.ADD_USER, payload: user });
  } catch (err) {
    console.log(err);
  }
};

export const deleteUser = (_id: string) => async (
  dispatch: Dispatch<Action>
) => {
  try {
    API.deleteUser(_id);
    dispatch({
      type: ActionTypes.DELETE_USER,
      payload: _id,
    });
  } catch (err) {
    console.log(err);
  }
};

export const updateUser = (user: IUser) => async (
  dispatch: Dispatch<Action>
) => {
  try {
    API.updateUser(user);
    dispatch({
      type: ActionTypes.UPDATE_USER,
      payload: user,
    });
  } catch (err) {
    console.log(err);
  }
};

export const bookmarkPost = (post: IPost) => async (
  dispatch: Dispatch<Action>
) => {
  try {
    dispatch({
      type: ActionTypes.BOOKMARK_POST,
      payload: post,
    });
  } catch (err) {
    console.log(err);
  }
};

export const getTopics = () => async (dispatch: Dispatch<Action>) => {
  dispatch({ type: ActionTypes.FETCH_TOPICS });

  try {
    const data = await API.getTopics();
    dispatch({
      type: ActionTypes.FETCH_TOPICS_SUCCESS,
      payload: data,
    });
  } catch (err) {
    dispatch({
      type: ActionTypes.FETCH_TOPICS_FAIL,
      payload: err,
    });
  }
};
