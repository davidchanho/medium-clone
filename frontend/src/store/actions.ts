import { ActionTypes } from ".";
import { IComment, IItem, IPost, IPublication, ITopic, IUser } from "../types";

export interface LoadingItemsAction {
  type: ActionTypes.LOADING_ITEMS;
}

export interface ErrorItemsAction {
  type: ActionTypes.ERROR_ITEMS;
  name: string;
  payload: string;
}

export interface FetchItemsAction {
  type: ActionTypes.FETCH_ITEMS;
  name: string;
  payload: IItem[];
}

export interface FetchItemAction {
  type: ActionTypes.FETCH_ITEM;
  name: string;
  payload: IItem;
}

export interface AddItemAction {
  type: ActionTypes.ADD_ITEM;
  name: string;
  payload: IItem;
}

export interface DeleteItemAction {
  type: ActionTypes.DELETE_ITEM;
  name: string;
  payload: string;
}

export interface UpdateItemAction {
  type: ActionTypes.UPDATE_ITEM;
  name: string;
  payload: IItem;
}

export interface LoadingCommentsAction {
  type: ActionTypes.LOADING_COMMENTS;
}

export interface ErrorCommentsAction {
  type: ActionTypes.ERROR_COMMENTS;
  payload: string;
}

export interface FetchCommentsAction {
  type: ActionTypes.FETCH_COMMENTS;
  payload: IComment[];
}

export interface FetchCommentAction {
  type: ActionTypes.FETCH_COMMENT;
  payload: IComment;
}

export interface AddCommentAction {
  type: ActionTypes.ADD_COMMENT;
  payload: IComment;
}

export interface DeleteCommentAction {
  type: ActionTypes.DELETE_COMMENT;
  payload: string;
}

export interface UpdateCommentAction {
  type: ActionTypes.UPDATE_COMMENT;
  payload: IComment;
}

export interface LoadingPostsAction {
  type: ActionTypes.LOADING_POSTS;
}

export interface FetchPostsAction {
  type: ActionTypes.FETCH_POSTS;
  payload: IPost[];
}

export interface ErrorPostsAction {
  type: ActionTypes.ERROR_POSTS;
  payload: string;
}

export interface FetchPostAction {
  type: ActionTypes.FETCH_POST;
  payload: IPost;
}

export interface AddPostAction {
  type: ActionTypes.ADD_POST;
  payload: IPost;
}

export interface DeletePostAction {
  type: ActionTypes.DELETE_POST;
  payload: string;
}

export interface UpdatePostAction {
  type: ActionTypes.UPDATE_POST;
  payload: IPost;
}

export interface LoadingPublicationsAction {
  type: ActionTypes.LOADING_PUBLICATIONS;
}

export interface FetchPublicationsAction {
  type: ActionTypes.FETCH_PUBLICATIONS;
  payload: IPublication[];
}

export interface ErrorPublicationsAction {
  type: ActionTypes.ERROR_PUBLICATIONS;
  payload: string;
}

export interface FetchPublicationAction {
  type: ActionTypes.FETCH_PUBLICATION;
  payload: IPublication;
}

export interface LoadingTopicsAction {
  type: ActionTypes.LOADING_TOPICS;
}

export interface FetchTopicsAction {
  type: ActionTypes.FETCH_TOPICS;
  payload: ITopic[];
}

export interface ErrorTopicsAction {
  type: ActionTypes.ERROR_TOPICS;
  payload: string;
}

export interface LoadingUsersAction {
  type: ActionTypes.LOADING_USERS;
}

export interface FetchUsersAction {
  type: ActionTypes.FETCH_USERS;
  payload: IUser[];
}

export interface ErrorUsersAction {
  type: ActionTypes.ERROR_USERS;
  payload: string;
}

export interface FetchUserAction {
  type: ActionTypes.FETCH_USER;
  payload: IUser;
}

export interface AddUserAction {
  type: ActionTypes.ADD_USER;
  payload: IUser;
}

export interface DeleteUserAction {
  type: ActionTypes.DELETE_USER;
  payload: string;
}

export interface UpdateUserAction {
  type: ActionTypes.UPDATE_USER;
  payload: IUser;
}

export interface BookmarkPostAction {
  type: ActionTypes.BOOKMARK_POST;
  payload: IPost;
}

export type Action =
  | FetchItemsAction
  | ErrorItemsAction
  | LoadingItemsAction
  | FetchItemAction
  | AddItemAction
  | DeleteItemAction
  | UpdateItemAction
  | FetchCommentsAction
  | ErrorCommentsAction
  | LoadingCommentsAction
  | FetchCommentAction
  | AddCommentAction
  | DeleteCommentAction
  | UpdateCommentAction
  | FetchPostsAction
  | ErrorPostsAction
  | LoadingPostsAction
  | FetchPostAction
  | AddPostAction
  | DeletePostAction
  | UpdatePostAction
  | FetchPublicationsAction
  | ErrorPublicationsAction
  | LoadingPublicationsAction
  | FetchPublicationAction
  | FetchTopicsAction
  | ErrorTopicsAction
  | LoadingTopicsAction
  | FetchUsersAction
  | ErrorUsersAction
  | LoadingUsersAction
  | FetchUserAction
  | AddUserAction
  | DeleteUserAction
  | UpdateUserAction
  | BookmarkPostAction;
