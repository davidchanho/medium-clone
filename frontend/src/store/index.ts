export * as actionCreators from "./actionCreators";
export * from "./actions";
export * from "./actionTypes";
export * from "./api";
export * from "./reducers";
export * from "./selectors";
export * from "./store";
