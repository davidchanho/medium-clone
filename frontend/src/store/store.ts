import localForage from "localforage";
import { applyMiddleware, createStore } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import { createLogger } from "redux-logger";
import { persistReducer, persistStore } from "redux-persist";
import thunk from "redux-thunk";
import reducers from "./reducers";

// localForage
localForage.config({
  driver: localForage.INDEXEDDB,
});

// redux-perist
const persistConfig = {
  key: "medium",
  storage: localForage,
  whitelist: ["posts"],
};
const persistedReducer = persistReducer(persistConfig, reducers);

// middleware
const logger = createLogger();
const middleware = [thunk, logger];

export const store = createStore(
  persistedReducer,
  {},
  composeWithDevTools(applyMiddleware(...middleware))
);

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
