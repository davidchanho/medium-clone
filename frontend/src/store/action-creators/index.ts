import { Dispatch } from "redux";
import db from "../../api";
import { ActionTypes } from "../action-types";
import { Action } from "../actions";
import { IPost, IPublication } from "../reducers/posts";

export const fetchPosts = () => async (dispatch: Dispatch<Action>) => {
  dispatch({ type: ActionTypes.FETCH_POSTS });

  try {
    const data = await db.fetchPosts();
    dispatch({
      type: ActionTypes.FETCH_POSTS_SUCCESS,
      payload: data,
    });
  } catch (err) {
    dispatch({
      type: ActionTypes.FETCH_POSTS_FAIL,
      payload: err,
    });
  }
};

export const fetchPost = (_id: string) => async (
  dispatch: Dispatch<Action>
) => {
  dispatch({ type: ActionTypes.FETCH_POST });

  try {
    const data = await db.fetchPost(_id);
    dispatch({
      type: ActionTypes.FETCH_POST_SUCCESS,
      payload: data,
    });
  } catch (err) {
    dispatch({
      type: ActionTypes.FETCH_POST_FAIL,
      payload: err,
    });
  }
};

export const addPost = (post: IPost) => async (dispatch: Dispatch<Action>) => {
  try {
    db.addPost(post);
    dispatch({ type: ActionTypes.ADD_POST, payload: post });
  } catch (err) {
    console.log(err);
  }
};

export const deletePost = (_id: string) => async (
  dispatch: Dispatch<Action>
) => {
  try {
    db.deletePost(_id);
    dispatch({
      type: ActionTypes.DELETE_POST,
      payload: _id,
    });
  } catch (err) {
    console.log(err);
  }
};

export const updatePost = (post: IPost) => async (
  dispatch: Dispatch<Action>
) => {
  try {
    db.updatePost(post);
    dispatch({
      type: ActionTypes.UPDATE_POST,
      payload: post,
    });
  } catch (err) {
    console.log(err);
  }
};

export const fetchPublications = () => async (dispatch: Dispatch<Action>) => {
  dispatch({ type: ActionTypes.FETCH_PUBLICATIONS });

  try {
    const data = await db.fetchPublications();
    dispatch({
      type: ActionTypes.FETCH_PUBLICATIONS_SUCCESS,
      payload: data,
    });
  } catch (err) {
    dispatch({
      type: ActionTypes.FETCH_PUBLICATIONS_FAIL,
      payload: err,
    });
  }
};

export const fetchPublication = (_id: string) => async (
  dispatch: Dispatch<Action>
) => {
  dispatch({ type: ActionTypes.FETCH_PUBLICATION });

  try {
    const data = await db.fetchPublication(_id);
    dispatch({
      type: ActionTypes.FETCH_PUBLICATION_SUCCESS,
      payload: data,
    });
  } catch (err) {
    dispatch({
      type: ActionTypes.FETCH_PUBLICATION_FAIL,
      payload: err,
    });
  }
};

export const addPublication = (publication: IPublication) => async (dispatch: Dispatch<Action>) => {
  try {
    db.addPublication(publication);
    dispatch({ type: ActionTypes.ADD_PUBLICATION, payload: publication });
  } catch (err) {
    console.log(err);
  }
};

export const deletePublication = (_id: string) => async (
  dispatch: Dispatch<Action>
) => {
  try {
    db.deletePublication(_id);
    dispatch({
      type: ActionTypes.DELETE_PUBLICATION,
      payload: _id,
    });
  } catch (err) {
    console.log(err);
  }
};

export const updatePublication = (publication: IPublication) => async (
  dispatch: Dispatch<Action>
) => {
  try {
    db.updatePublication(publication);
    dispatch({
      type: ActionTypes.UPDATE_PUBLICATION,
      payload: publication,
    });
  } catch (err) {
    console.log(err);
  }
};
