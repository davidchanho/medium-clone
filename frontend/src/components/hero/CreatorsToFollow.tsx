import React from "react";
import { ListGroup } from "react-bootstrap";
import { useSelector } from "react-redux";
import { userSelectors } from "../../store";
import { IUser } from "../../store/users/types";
import FellowCreator from "../side-bar/FollowCreator";

function CreatorsToFollow() {
  const { users, loading, error } = useSelector(userSelectors);

  if (loading) {
    return <h2>Loading...</h2>;
  }

  if (error) {
    return <h2>{error}</h2>;
  }

  if (!users) {
    return null;
  }

  const renderCreators = () => {
    return (
      <ListGroup>
        {users.slice(0, 2).map((user: IUser) => {
          return (
            <FellowCreator
              imgStyles="rounded-circle"
              key={`follow-creators-${user?._id}`}
              user={user}
            />
          );
        })}
      </ListGroup>
    );
  };

  return (
    <div>
      <p className="text-uppercase font-weight-bold">Creators to follow</p>{" "}
      {renderCreators()}
    </div>
  );
}

export default CreatorsToFollow;
