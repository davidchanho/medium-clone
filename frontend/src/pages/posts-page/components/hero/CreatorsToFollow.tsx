import React from "react";
import { useSelector } from "react-redux";
import { userSelectors } from "../../../../store";
import { IUser } from "../../../../types";
import Creator from "./Creator";

function CreatorsToFollow() {
  const { users } = useSelector(userSelectors);

  return (
    <div>
      <p>Creators to follow</p>{" "}
      <ul>
        {users.slice(0, 2).map((user: IUser) => {
          return <Creator key={`creators-to-follow-${user._id}`} {...user} />;
        })}
      </ul>
    </div>
  );
}

export default CreatorsToFollow;
