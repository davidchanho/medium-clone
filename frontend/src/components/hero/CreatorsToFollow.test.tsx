import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import { store } from "../../store";
import CreatorsToFollow from "./CreatorsToFollow";

describe("CreatorsToFollow", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <CreatorsToFollow />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
