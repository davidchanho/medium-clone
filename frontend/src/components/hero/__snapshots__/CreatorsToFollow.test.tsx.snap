// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatorsToFollow renders correctly 1`] = `
<div
  className="mb-3"
>
  <p
    className="text-uppercase font-weight-bold mb-2"
  >
    Creators to follow
  </p>
   
  <div
    className="list-group"
    onKeyDown={[Function]}
  />
</div>
`;
